### PR TITLE
Better envelope intersection

### PR DESCRIPTION
**Pull request by multun**
- Introduce a more versatile overlay envelope builder, which handles intersections with arbitrary constraints.
- Remove the old OverlayEnvelopePartBuilder

Depends on #604 
Required for hardening MARECO

### CODE REVIEW

File: core/src/main/java/fr/sncf/osrd/envelope/Envelope.java [hunk 1]
```diff
@@ -159,6 +148,20 @@ public int findEnvelopePartIndexRight(double position) {
         return -1;
     }
 
+    /** Returns the last envelope part index along the given direction. */
+    public int findEnvelopePartIndexLeftDir(double position, double direction) {
+        if (direction > 0)
+            return findEnvelopePartIndexLeft(position);
+        return findEnvelopePartIndexRight(position);
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/envelope/Envelope.java:155]:
```suggestion
        return findEnvelopePartIndexRightDir(position, -direction);
```

File: core/src/main/java/fr/sncf/osrd/envelope/NullEnvelopePartConsumer.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package fr.sncf.osrd.envelope;
+
+public class NullEnvelopePartConsumer implements EnvelopePartConsumer {
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/envelope/NullEnvelopePartConsumer.java:3]: We could move this in `tests` package since it's not used by the `main`.

**Review by flomonster — COMMENTED** [review]
Big PR, I love the new envelope part constraint systems.

File: core/src/main/java/fr/sncf/osrd/envelope/NullEnvelopePartConsumer.java (hunk 2 above)
Comment (multun) [reply core/src/main/java/fr/sncf/osrd/envelope/NullEnvelopePartConsumer.java:3]: It's soon to be used by the new Mareco

File: core/src/main/java/fr/sncf/osrd/envelope/Envelope.java (hunk 1 above)
Comment (multun) [reply core/src/main/java/fr/sncf/osrd/envelope/Envelope.java:155]: I'd rather keep it symmetric

**Pull request merged**